ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/altus/GrpcMinaSshdManagementClient.java

```diff
@@ -0,0 +1,214 @@
+        switch (service.getStatus()) {
+            case STARTED:
+                if (service.getReady()) {
+                    return Optional.of(service);
+                }
+                break;
+            case FAILED:
+                throw new CcmException(String.format("Failed while trying to %s", actionDescription), false);
+            default:
+                break;
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Polls trying to find a service which is ready and has a status of {@code STARTED}.
+     *
+     * @param listAction               the action to list the minasshd services (must return at most 1 service)
+     * @param actionDescription        the action description, for logging and exception messages
+     * @param waitUntilTime            the latest time to wait to perform the action
+     * @param pollingIntervalMillis    the polling interval in milliseconds
+     * @param timeoutExceptionSupplier a supplier of exceptions for when all attempts fail
+     * @return an optional wrapping a service that is ready and has a status of {@code STARTED}, or an empty optional if there are no pending services
+     * @throws InterruptedException if the thread is interrupted
+     * @throws CcmException         if the operation times out with services still pending, or if a non-transient error occurs
+     */
+    private Optional<MinaSshdService> awaitValidMinaSshdService(Callable<List<MinaSshdService>> listAction, String actionDescription,
+            ZonedDateTime waitUntilTime, int pollingIntervalMillis,
+            Supplier<CcmException> timeoutExceptionSupplier)
+            throws InterruptedException, CcmException {
+        long waitUntilTimeMillis = waitUntilTime.toInstant().toEpochMilli();
+
+        while (true) {
+
+            // Call listMinaSshdServices, with retries in case of transient failures.
+            List<MinaSshdService> minaSshdServices = RetryUtil.performWithRetries(
+                    listAction, actionDescription, waitUntilTime, pollingIntervalMillis,
+                    CcmException.class, timeoutExceptionSupplier,
+                    LOGGER);
+
+            MinaSshdService service;
+            switch (minaSshdServices.size()) {
+                case 0:
+                    return Optional.empty();
+                case 1:
+                    service = minaSshdServices.get(0);
+                    break;
+                default:
+                    throw new IllegalStateException("listAction returned multiple services");
+            }
+
+            switch (service.getStatus()) {
+                case STARTED:
+                    if (service.getReady()) {
+                        return Optional.of(service);
+                    }
+                    break;
+                case FAILED:
+                    return Optional.empty();
```

Comment:
shouldn't this throw CCMException here? This feels inconsistent with getValidMinaSshdService

Reply:
will check.